File: esw-utils/src/test/scala/esw/highlevel/dsl/TimeServiceDslTest.scala
package esw.highlevel.dsl

import java.time.Duration

import akka.actor.Scheduler
import csw.time.core.models.{TMTTime, UTCTime}
import csw.time.scheduler.TimeServiceSchedulerFactory
import csw.time.scheduler.api.{Cancellable, TimeServiceScheduler}
import esw.ocs.api.BaseTestSuite
import esw.ocs.macros.StrandEc
import org.mockito.ArgumentMatchers.{any, eq => argsEq}

import scala.concurrent.ExecutionContext
import scala.concurrent.duration.{DurationInt, FiniteDuration}

class TimeServiceDslTest extends BaseTestSuite with TimeServiceDsl {

  implicit val scheduler: Scheduler         = mock[Scheduler]
  private implicit val ec: ExecutionContext = StrandEc().ec

  val cancellable: Cancellable                                 = mock[Cancellable]
  val timeServiceScheduler: TimeServiceScheduler               = mock[TimeServiceScheduler]
  val timeServiceSchedulerFactory: TimeServiceSchedulerFactory = mock[TimeServiceSchedulerFactory]
  val task: Unit                                               = {}

  "ScheduleOnce" must {
    "delegate to schedule tasks for given start time" in {
      val startTime = UTCTime.now()

      when(timeServiceSchedulerFactory.make()).thenReturn(timeServiceScheduler)
      when(timeServiceScheduler.scheduleOnce(startTime)(task)).thenReturn(cancellable)

      val actualCancellable = scheduleOnce(startTime)(task)

      verify(timeServiceScheduler).scheduleOnce(startTime)(task)
      actualCancellable shouldBe cancellable
    }
  }

  "schedulePeriodically" must {
    "delegate to schedule task for given interval with default time" in {
      val task: Unit               = {}
      val interval: FiniteDuration = 5.seconds

      when(timeServiceSchedulerFactory.make()).thenReturn(timeServiceScheduler)

      schedulePeriodically(interval)(task)

      verify(timeServiceScheduler).schedulePeriodically(any[TMTTime], argsEq(Duration.ofNanos(interval.toNanos)))(argsEq(task))
    }

    "delegate to schedule task for given interval for given time" in {
      val task: Unit               = {}
      val interval: FiniteDuration = 5.seconds
      val startTime                = UTCTime.now()

      when(timeServiceSchedulerFactory.make()).thenReturn(timeServiceScheduler)

      schedulePeriodically(interval, startTime)(task)

      verify(timeServiceScheduler).schedulePeriodically(argsEq(startTime), argsEq(Duration.ofNanos(interval.toNanos)))(
        argsEq(task)
      )
    }
  }
}

File: script-dsl/src/test/kotlin/esw/ocs/dsl/highlevel/TimeServiceDslTest.kt
package esw.ocs.dsl.highlevel

import csw.time.core.models.UTCTime
import csw.time.scheduler.api.Cancellable
import csw.time.scheduler.api.TimeServiceScheduler
import io.kotlintest.matchers.beGreaterThanOrEqualTo
import io.kotlintest.matchers.beLessThanOrEqualTo
import io.kotlintest.shouldBe
import io.kotlintest.specs.WordSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import java.util.concurrent.TimeUnit
import kotlin.coroutines.CoroutineContext
import kotlin.time.*
import kotlinx.coroutines.Job
import scala.concurrent.duration.FiniteDuration

class TimeServiceDslTest : WordSpec() {
    class Mocks {
        val scheduler = mockk<TimeServiceScheduler>()
        val cancellable = mockk<Cancellable>()

        val startTime: UTCTime = UTCTime.now()
        val duration: Duration = 10.seconds
        val jDuration = duration.toJavaDuration()
        val context = Job()

        val timeServiceDsl = object : TimeServiceDsl {
            override val timeServiceScheduler: TimeServiceScheduler = scheduler
            override val coroutineContext: CoroutineContext = context
        }
    }

    abstract class TestTimeServiceDsl(mocks: Mocks) : TimeServiceDsl {
        override val timeServiceScheduler: TimeServiceScheduler = mocks.timeServiceDsl.timeServiceScheduler
        override val coroutineContext: CoroutineContext = mocks.timeServiceDsl.coroutineContext
    }

    private fun checkWithIn(expected: Duration, result: Duration, tolerance: Duration) {
        val diff = expected - result
        diff shouldBe beLessThanOrEqualTo(tolerance)
        diff shouldBe beGreaterThanOrEqualTo(-tolerance)
    }

    init {
        "TimeServiceDsl" should {
            "scheduleOnce should delegate to timeServiceScheduler.scheduleOnce | ESW-122" {
                with(Mocks()) {
                    every { scheduler.scheduleOnce(startTime, any<Runnable>()) }.answers { cancellable }

                    timeServiceDsl.scheduleOnce(startTime, mockk()) shouldBe cancellable

                    verify { scheduler.scheduleOnce(startTime, any<Runnable>()) }
                }
            }

            "schedulePeriodically should delegate to timeServiceScheduler.schedulePeriodically | ESW-122" {
                with(Mocks()) {
                    every {
                        scheduler.schedulePeriodically(startTime, jDuration, any<Runnable>())
                    }.answers { cancellable }

                    timeServiceDsl.schedulePeriodically(startTime, duration, mockk()) shouldBe cancellable

                    verify { scheduler.schedulePeriodically(startTime, jDuration, any<Runnable>()) }
                }
            }

            "offsetFromNow should give offset between current time and provided instance of the time" {
                with(Mocks()) {
                    val dsl = object : TestTimeServiceDsl(this) {
                        fun testOffset() {
                            val offset: Duration = UTCTime.after(FiniteDuration(1, TimeUnit.SECONDS)).offsetFromNow()
                            checkWithIn(expected = 1.seconds, result = offset, tolerance = 2.milliseconds)
                        }
                    }

                    dsl.testOffset()
                }
            }

            "utcTimeAfter should give UTC time of after given duration from now" {
                with(Mocks()) {
                    val dsl = object : TestTimeServiceDsl(this) {
                        fun testUtcAfter() {
                            val offset: Duration = utcTimeAfter(1.seconds).offsetFromNow()
                            checkWithIn(expected = 1.seconds, result = offset, tolerance = 2.milliseconds)
                        }
                    }

                    dsl.testUtcAfter()
                }
            }

            "taiTimeAfter should give time TAI of after given duration from now" {
                with(Mocks()) {
                    val dsl = object : TestTimeServiceDsl(this) {
                        fun testTaiAfter() {
                            val offset: Duration = taiTimeAfter(1.seconds).offsetFromNow()
                            checkWithIn(expected = 1.seconds, result = offset, tolerance = 2.milliseconds)
                        }
                    }

                    dsl.testTaiAfter()
                }
            }
        }
    }
}
